src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/DomainEventService.kt

```diff
@@ -0,0 +1,128 @@
+import com.amazonaws.services.sns.model.PublishRequest
+import com.fasterxml.jackson.databind.ObjectMapper
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.events.cas2.model.Cas2ApplicationSubmittedEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.api.events.cas2.model.Cas2Event
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventEntity
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.DomainEventType
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.DomainEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEvent
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventAdditionalInformation
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReference
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.model.domainevent.SnsEventPersonReferenceCollection
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
+import uk.gov.justice.hmpps.sqs.MissingTopicException
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.util.UUID
+import javax.transaction.Transactional
+import kotlin.reflect.KClass
+
+@Service(
+  "uk.gov.justice.digital.hmpps.approvedpremisesapi.service.cas2.DomainEventService",
+)
+class DomainEventService(
+  private val objectMapper: ObjectMapper,
+  private val domainEventRepository: DomainEventRepository,
+  private val hmppsQueueService: HmppsQueueService,
+  @Value("\${domain-events.cas2.emit-enabled}") private val emitDomainEventsEnabled: Boolean,
+  @Value("\${url-templates.api.cas2.application-submitted-event-detail}") private val cas2ApplicationSubmittedDetailUrlTemplate: String,
+) {
+  private val log = LoggerFactory.getLogger(this::class.java)
+
+  private val domainTopic by lazy {
+    hmppsQueueService.findByTopicId("domainevents")
+      ?: throw MissingTopicException("domainevents not found")
+  }
+
+  fun getCas2ApplicationSubmittedDomainEvent(id: UUID) = get<Cas2ApplicationSubmittedEvent>(id)
+
+  private inline fun <reified T : Cas2Event> get(id: UUID): DomainEvent<T>? {
+    val domainEventEntity = domainEventRepository.findByIdOrNull(id) ?: return null
+
+    val data = when {
+      enumTypeFromDataType(T::class) == domainEventEntity.type ->
+        objectMapper.readValue(domainEventEntity.data, T::class.java)
+      else -> throw RuntimeException("Unsupported DomainEventData type ${T::class.qualifiedName}/${domainEventEntity.type.name}")
+    }
+    return DomainEvent(
+      id = domainEventEntity.id,
+      applicationId = domainEventEntity.applicationId,
+      bookingId = domainEventEntity.bookingId,
```

Comment:
hmm, this `bookingId`... As we don't have bookings would it be clearer to explicitly set this to `null`?

Reply:
ah yeah, thanks this one slipped through 🐟